### ISSUE TITLE
Disabling ICC doesn't block Raw Ethernet packets between containers

### ISSUE THREAD

**Issue by brthor** [moby/moby#36355]
**Description**

When ICC is disabled on the daemon via `--icc` or on an individual network, it is expected that **ALL** network communications are blocked. 

I have demonstrated that Raw Ethernet frames can still be transmitted between containers when ICC is disabled.

**Steps to reproduce the issue:**

I've created a separate repository solely for reproduction as a few files need to be hosted:
https://github.com/brthor/docker-layer2-icc

**Describe the results you received:**
I was able to send Raw ethernet packets between containers on a network bridge with ICC disabled.

**Describe the results you expected:**
I should not have been able to send Raw ethernet packets between containers on a network bridge with ICC disabled.

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `docker version`:**

```
Client:
 Version:      17.09.0-ce-rc3
 API version:  1.32
 Go version:   go1.8.3
 Git commit:   2357fb2
 Built:        Thu Sep 21 02:31:18 2017
 OS/Arch:      darwin/amd64

Server:
 Version:      17.09.0-ce-rc3
 API version:  1.32 (minimum version 1.12)
 Go version:   go1.8.3
 Git commit:   2357fb2
 Built:        Thu Sep 21 02:36:52 2017
 OS/Arch:      linux/amd64
 Experimental: true
```

**Output of `docker info`:**

```
Containers: 320
 Running: 1
 Paused: 0
 Stopped: 319
Images: 2547
Server Version: 17.09.0-ce-rc3
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 3099
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host ipvlan macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 06b9cb35161009dcb7123345749fef02f7cea8e0
runc version: 3f2f8b84a77f73d38244dd690525642a72156c64
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.9.49-moby
Operating System: Alpine Linux v3.5
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 1.952GiB
Name: moby
ID: FFPF:MCST:3YKS:C6WH:XMZJ:YS5R:QDGW:GNAT:EFJ2:PIJP:UK5X:OOPX
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 31
 Goroutines: 56
 System Time: 2018-02-20T03:28:39.232321839Z
 EventsListeners: 1
No Proxy: *.local, 169.254/16
Registry: https://index.docker.io/v1/
Experimental: true
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

Using Docker for Mac.

**Comment by brthor** [comment]
I'm going to make a stab at working around this, but pointers where to start would be appreciated.

**Comment by brthor** [comment]
ping @TheJetzah

**Comment by brthor** [comment]
Second repro on CentOS 7.3 w/ SELinux enabled on latest edge version `18.02.0-ce`

Unlikely to be a platform specific issue.

Repro steps simplified to clone and run.

**Comment by justincormack** [comment]
You should always run with `--cap-drop CAP_NET_RAW` if you want to block raw traffic. This is the only capability on the default capability set that really should not be there as it does allow some extra abilities, however people expect to be able to run `ping` form containers and ping sockets (which are restricted to only icmp) have not reached many Linux distros (only Alpine last time I checked), and still require per-container sysctl setup to use so are pretty inconvenient.

(We could provide a config to reduce the default capability set, I have considered this before).

**Comment by brthor** [comment]
Indeed `--cap-drop CAP_NET_RAW` does prevent this, but doesn't alleviate the mismatch in expectation this bug causes. There is still expectation that disabling inter-container communications would extend to the data passed over raw sockets. 

There may be applications for which raw sockets are desired, and proper networking sandboxing is also desired. In a realistic sense, it seems that it could cause problems for unaware developers relying on disabling ICC.

**Comment by justincormack** [comment]
I don't believe you can have reliable network sandboxing at all without disabling raw sockets, on a bridged network (overlay networks differ), as you can arp spoof and listen for other traffic. If you put each container on a different bridge you may be able to.

**Comment by brthor** [comment]
I was able to block the communication without dropping `CAP_NET_RAW` using multiple ebtables rules.

`ebtables` is apparently an application similar to ip tables which filters at layer 2. The rule syntax is very similar.

The rules I added were:
```
sudo ebtables -A FORWARD -d 02:42:ac:13:00:02 -s 02:42:ac:13:00:03 -j DROP
sudo ebtables -A FORWARD -d  02:42:ac:13:00:03 -s 02:42:ac:13:00:02 -j DROP
```

Both MAC addresses above represent containers on the same bridge network.
Two rules for every pair of containers on the network seemingly would need to be added unless there is a way to combine rules.

**Comment by brthor** [comment]
@justincormack 

I agree with you, that reliable sandboxing while retaining raw sockets will be difficult. Believe me I am already blocking them in my deployments, and with a custom seccomp profile, and will drop it from the SELinux policy as soon as I figure out how to load one besides `container-selinux`.

If capabilities were a magic bullet though, tools like seccomp and SELinux wouldn't exist. My intentions with desiring some default firewall between containers at layer 2 is rooted in the same philosophy. The lack of perfect security shouldn't prevent us from making improvements where possible.

ARP spoofing is the obvious elephant in the room with this bug, although I didn't want to point it out. The `ebtables` [website](http://ebtables.netfilter.org) makes some claims that it can have "anti-spoofing" filters. I plan on investigating that soon, and will report on what I find.

When I was writing the code for this repro, I was building out a test, so I can reliably determine that network communications are sandboxed at the ip layer, but also at layer 2. I use such tests to ensure the security of my environment between updates of docker and other components. 

The `ebtables` based fix is simple enough but presents an issue in that it will create a new system dependency for docker. I know this can cause a lot of complexity, especially with how many platforms docker supports. How are conversations around this type of thing usually conducted in docker? 

FYI: I tried creating `iptables` rule but filtering on the destination mac doesn't seem to be possible.

**Comment by AkihiroSuda** [comment]
> The ebtables based fix is simple enough but presents an issue in that it will create a new system dependency for docker, and may have unexplored side effects. 

If we can enable ebtables only when `--ebtables=true` is specified, I think it is fine.


BTW [nftables](https://wiki.nftables.org/wiki-nftables/index.php/Nftables_families#netdev) seems to work for L2, can we use that instead?

Also, I heard nftables can replace iptables and it is faster than iptables for complex rules.
So I think we can use nftables for both L2 and L3.

**Comment by justincormack** [comment]
You can't do arp spoofing if you disable `CAP_NET_RAW` either. I don't understand what you are saying about "If capabilities were a magic bullet though, tools like seccomp and SELinux wouldn't exist." - capabilities have some issues in that they are very coarse grained but this one is exactly what you need, and is designed for this. It is actually very difficult to block raw sockets with seccomp, I don't recommend trying it. SELinux is not the right tool for this either. Alternatively, you can just not run containers as the root user, which is also recommended, in which case they will not be able to use raw sockets either.

**Comment by thaJeztah** [comment]
> BTW nftables seems to work for L2, can we use that instead?

There's an open (but quite old) issue about using nftables; https://github.com/moby/moby/issues/26824

**Comment by justincormack** [comment]
At this point `nftables` is largely a failed experiment with no widespread use.

**Comment by brthor** [comment]
No experience with `nftables` but it seems like a larger change with more risk.
